fvm/programs.go

```diff
@@ -0,0 +1,92 @@
+package fvm
+
+import (
+	"sync"
+
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+)
+
+type ProgramEntry struct {
+	Location common.Location
+	Program  *interpreter.Program
+}
+
+type Programs struct {
+	lock     sync.RWMutex
+	parent   *Programs
+	programs map[common.LocationID]ProgramEntry
+}
+
+func NewPrograms(parent *Programs) *Programs {
+	return &Programs{
+		parent:   parent,
+		programs: map[common.LocationID]ProgramEntry{},
+	}
+}
+
+func (p *Programs) Get(locationID common.LocationID) *interpreter.Program {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	// First check if the program is available here
+	if programEntry, ok := p.programs[locationID]; ok {
+		return programEntry.Program
+	}
+
+	// Second, check if the program is available in the parent (if any)
+
+	if p.parent != nil {
+		return p.parent.Get(locationID)
+	}
+
+	// The program is neither available here nor in the parent
+
+	return nil
+}
+
+func (p *Programs) Set(location common.Location, program *interpreter.Program) {
```

Comment:
Any value to add the logic here to when we set an already existing value we purge the map ?